### PR TITLE
Remove the SoftLimiter and related buffers from GUS

### PR DESCRIPTION
**Pull request by kcgen**
In preparation for managing the levels in the mixer!

### CODE REVIEW

**Review by Burrito78 — APPROVED** [review]

File: src/hardware/gus.cpp [hunk 1]
```diff
@@ -271,10 +272,17 @@ class Gus {
...
 	// Struct and pointer members
 	VoiceIrq voice_irq = {};
-	SoftLimiter soft_limiter;
 	Voice *target_voice = nullptr;
 	DmaChannel *dma_channel = nullptr;
 	mixer_channel_t audio_channel = nullptr;
+	AudioFrame accumulator_scalar = {};
+
+	// Playback related
+	double last_render_time = 0.0;
```
Comment (johnnovak) [inline src/hardware/gus.cpp:281]: last_render_time_ms? (for consistency with the other vars)

File: src/hardware/gus.cpp [hunk 2]
```diff
@@ -628,44 +623,112 @@ void Gus::ActivateVoices(uint8_t requested_voices)
...
 		}
-		soft_limiter.Process(render_buffer, frames, play_buffer);
-		audio_channel->AddSamples_s16(frames, play_buffer.data());
-		CheckVoiceIrq();
-		generated_frames += frames;
+		accumulator.left *= accumulator_scalar.left;
+		accumulator.right *= accumulator_scalar.right;
+	}
+	        CheckVoiceIrq();
+	        return accumulator;
```
Comment (johnnovak) [inline src/hardware/gus.cpp:670]: indentation

File: src/hardware/gus.cpp [hunk 3]
```diff
@@ -628,44 +623,112 @@ void Gus::ActivateVoices(uint8_t requested_voices)
...
+		while (requested_frames--) {
+			const auto frame = RenderFrame();
+			audio_channel->AddSamples_sfloat(1, &frame[0]);
+		}
+	}
+	// Pause the card if it hasn't been written to for 3 seconds
+	constexpr uint16_t three_seconds_of_ticks = 3 * 1000;
+	if (++unused_for_ms > three_seconds_of_ticks) {
+		audio_channel->Enable(false);
+		LOG_MSG("sleeping GUS!");
```
Comment (johnnovak) [inline src/hardware/gus.cpp:731]: Nice descriptive variable names everywhere — I like it! 🎉

File: src/hardware/gus.cpp [hunk 4]
```diff
@@ -628,44 +623,112 @@ void Gus::ActivateVoices(uint8_t requested_voices)
...
+double Gus::ConvertFramesToMs(const int frames) const
+{
+	return frames / frame_rate_per_ms;
+}
+
+void Gus::AudioCallback(uint16_t requested_frames)
+{
+	assert(audio_channel);
+	while (requested_frames && fifo.size()) {
+		audio_channel->AddSamples_sfloat(1, &fifo.front()[0]);
```
Comment (johnnovak) [inline src/hardware/gus.cpp:715]: 🎉

File: src/hardware/gus.cpp [hunk 5]
```diff
@@ -628,44 +623,112 @@ void Gus::ActivateVoices(uint8_t requested_voices)
...
+		fifo.pop();
+		--requested_frames;
+	}
+
+	if (requested_frames) {
+		last_render_time += ConvertFramesToMs(requested_frames);
+		while (requested_frames--) {
+			const auto frame = RenderFrame();
+			audio_channel->AddSamples_sfloat(1, &frame[0]);
```
Comment (johnnovak) [inline src/hardware/gus.cpp:724]: 🥇

File: src/hardware/gus.cpp (hunk 3 above)
Comment (johnnovak) [reply src/hardware/gus.cpp:731]: Use same prefix here too? E.g. "GUS: Sleeping"

**Review by johnnovak — APPROVED** [review]
Great work & nice to see the float version of AddSamples being put to good use! 🎊

File: src/hardware/gus.cpp (hunk 3 above)
Comment (kcgen) [reply src/hardware/gus.cpp:731]: (thats for catching it! Will remove this sleep log..just a check to see if it's working)

File: src/hardware/gus.cpp [hunk 6]
```diff
@@ -628,44 +623,112 @@ void Gus::ActivateVoices(uint8_t requested_voices)
 		active_voices = requested_voices;
 		assert(active_voices <= voices.size());
 		active_voice_mask = 0xffffffffu >> (MAX_VOICES - active_voices);
-		playback_rate = static_cast<int>(
-		        round(1000000.0 / (1.619695497 * active_voices)));
-		audio_channel->SetSampleRate(playback_rate);
+
+		// Gravis' calculation to convert from number of active voices
+		// to playback frame rate. Ref: UltraSound Lowlevel ToolKit
+		// v2.22 (21 December 1994), pp. 3 of 113.
```
[the remark this answers is not in the document]
Comment (kcgen) [reply src/hardware/gus.cpp:629]: For anyone searching the tickets for Gravis Ultrasound Documentation:

[UltraSound Lowlevel ToolKit v2.22 (21 December 1994).pdf](https://github.com/dosbox-staging/dosbox-staging/files/8969393/UltraSound.Lowlevel.ToolKit.v2.22.21.December.1994.pdf)

File: src/hardware/gus.cpp (hunk 5 above)
Comment (kcgen) [reply src/hardware/gus.cpp:724]: It sure is nice passing through floats as-is, without any casting needed!

It's also really nice using `AudioFrame`, and I can see here that being able to pass a single `AudioFrame` (perhaps by reference?) or even an array/vector of them, would be quite a natural fit!

Soon enough :-)

**Pull request merged**